core/shared/src/main/scala/org/typelevel/idna4s/core/CodePoint.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2022 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package org.typelevel.idna4s.core
+
+import cats.kernel._
+import cats.Show
+import cats.syntax.all._
+
+/**
+ * A newtype for representing a Unicode code point.
+ *
+ * Unicode code points are a integral values from [0, 0x10FFF].
+ *
+ * This type is not (and should not) generally in critical path performance sensitive code. It
+ * is mostly useful for generating error information, testing, and for some performance
+ * insensitive methods which operate only a single code point. In particular for errors messages
+ * where we generally should not render arbitrary code points as they may influence the output
+ * text in ways we do not anticipate, such as bidi scripts or where the code point in isolation
+ * is not able to be rendered at all.
+ *
+ * @see
+ *   [[https://www.unicode.org/reports/tr36/ Unicode Security Considerations]]
+ */
+sealed abstract class CodePoint extends Serializable {
```

Comment:
I know it's not meant for performance but it does seem tempting to make this an `AnyVal` 😛

Reply:
Yeah, that makes sense here. I'll update it. I almost want to have a Scala 3 specific opaque type, but I don't think that makes sense as long as we cross support Scala 2.